File: aoc-py/solutions/day6.py
"""
Day 6: Guard Gallivant

https://adventofcode.com/2024/day/6
"""
__all__ = ('Day6',)

from typing import ClassVar

from ..solution import Solution

class Day6(Solution):
    NAME: ClassVar[str] = 'Guard Gallivant'

    def _find_start(self, grid: list[str]) -> tuple[int, int]:
        for i, row in enumerate(grid):
            for j, cell in enumerate(row):
                if cell == '^':
                    return i, j
        raise ValueError("No '^' character found in grid")

    def turn_right(self, dr: int, dc: int) -> tuple[int, int]:
        match (dr, dc):
            case (1, 0):
                return 0, -1
            case (-1, 0):
                return 0, 1
            case (0, 1):
                return 1, 0
            case (0, -1):
                return -1, 0
            case _:
                return 0, 0

    def part_one(self, inp: str) -> int:
        grid = inp.splitlines()
        n_rows = len(grid)
        n_cols = len(grid[0])

        row, col = self._find_start(grid)
        dr, dc = -1, 0

        seen = {(row, col)}

        while True:
            row += dr
            col += dc

            # guard has exited the area
            if row not in range(n_rows) or col not in range(n_cols):
                break

            if grid[row][col] == '#':
                # backtrack upon hitting obstacle
                row -= dr
                col -= dc

                # turn 90 degrees right
                match (dr, dc):
                    case (1, 0):
                        dr, dc = 0, -1
                    case (-1, 0):
                        dr, dc = 0, 1
                    case (0, 1):
                        dr, dc = 1, 0
                    case (0, -1):
                        dr, dc = -1, 0
            else:
                seen.add((row, col))

        return len(seen)

    def traverse_line(self, grid: list[str], set1: set, n_rows: int, n_cols: int, row: int, col: int, dr: int, dc: int) -> None:
        row += dr
        col += dc

        if row not in range(n_rows) or col not in range(n_cols) or grid[row][col] == '#':
            return

        set1.add((row, col, -dr, -dc))

        drr, dcc = self.turn_right(dr, dc)

        if row + drr in range(n_rows) and col + dcc in range(n_cols) and grid[row + drr][col + dcc] == '#':
            self.traverse_line(grid, set1, n_rows, n_cols, row, col, -drr, -dcc)
        else:
            self.traverse_line(grid, set1, n_rows, n_cols, row, col, dr, dc)

    def part_two(self, inp: str) -> int:
        grid = inp.splitlines()
        n_rows = len(grid)
        n_cols = len(grid[0])

        row, col = self._find_start(grid)
        dr, dc = -1, 0

        set1 = set()
        seen = [(row, col, dr, dc)]

        while True:
            row += dr
            col += dc

            # guard has exited the area
            if row not in range(n_rows) or col not in range(n_cols):
                break

            if grid[row][col] == '#':
                # backtrack upon hitting obstacle
                row -= dr
                col -= dc

                # turn 90 degrees right
                dr, dc = self.turn_right(dr, dc)

                self.traverse_line(grid, set1, n_rows, n_cols, row, col, dr, dc)
                self.traverse_line(grid, set1, n_rows, n_cols, row, col, -dr, -dc)
            else:
                seen.append((row, col, dr, dc))
            set1.add((row, col, dr, dc))
        total = 0
        for i, row in enumerate(grid):
            for j, cell in enumerate(row):
                if any(x[:2] == (i, j) for x in set1):
                    print('&', end='')
                else:
                    print(cell, end='')
            print()

        for i, (row, col, dr, dc) in enumerate(seen):
            drr, dcc = self.turn_right(dr, dc)
            if (row, col, drr, dcc) in set1 and not any(x[:2] == (row + dr, col + dc) for x in seen[:i]):
                print((row + dr, col + dc))
                total += 1
        return total

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        #assert p1 == 4696
        #assert p2 == 70478672

File: src/bin/day6.rs
//! Day 6: Guard Gallivant
//!
//! <https://adventofcode.com/2024/day/6>

use std::{
    collections::HashSet,
    fmt::Display,
    ops::Range,
};
use aoc_2024::{Solution, get_grid, find_start};

pub struct Day6;

impl Day6 {
    fn turn_right(dr: i8, dc: i8) -> (i8, i8) {
        match (dr, dc) {
            (1, 0) => (0, -1),
            (-1, 0) => (0, 1),
            (0, 1) => (1, 0),
            (0, -1) => (-1, 0),
            _ => unreachable!(),
        }
    }

    #[allow(clippy::cast_sign_loss)]
    fn get_path(
        grid: &[Vec<u8>],
        rows: &Range<usize>,
        cols: &Range<usize>,
        start @ (mut row, mut col): (usize, usize),
    ) -> HashSet<(usize, usize)> {
        let mut dr = -1;
        let mut dc = 0;

        let mut seen = HashSet::new();
        seen.insert(start);

        loop {
            row = row.wrapping_add(dr as usize);
            col = col.wrapping_add(dc as usize);

            if !rows.contains(&row) || !cols.contains(&col) {
                break seen;
            }

            if grid[row][col] == b'#' {
                row = row.wrapping_sub(dr as usize);
                col = col.wrapping_sub(dc as usize);

                (dr, dc) = Self::turn_right(dr, dc);
            } else {
                seen.insert((row, col));
            }
        }
    }
}

impl Solution for Day6 {
    const NAME: &'static str = "Guard Gallivant";

    fn part_one<T: Display>(&self, inp: T) -> Self::OutputP1 {
        let grid = get_grid(inp);
        let start = find_start(&grid, b'^').unwrap();

        let rows = 0..grid.len();
        let cols = 0..grid[0].len();

        Self::get_path(&grid, &rows, &cols, start).len()
    }

    #[allow(clippy::cast_sign_loss)]
    fn part_two<T: Display>(&self, inp: T) -> Self::OutputP2 {
        let grid = get_grid(inp);
        let start = find_start(&grid, b'^').unwrap();

        let rows = 0..grid.len();
        let cols = 0..grid[0].len();

        let mut total = 0;

        for obstacle in Self::get_path(&grid, &rows, &cols, start) {
            let (mut row, mut col) = start;

            let mut dr = -1;
            let mut dc = 0;

            let mut seen = HashSet::new();
            seen.insert((row, col, dr, dc));

            loop {
                row = row.wrapping_add(dr as usize);
                col = col.wrapping_add(dc as usize);

                if seen.contains(&(row, col, dr, dc)) {
                    total += 1;
                    break;
                }

                if !rows.contains(&row) || !cols.contains(&col) {
                    break;
                }

                if grid[row][col] == b'#' || (row, col) == obstacle {
                    row = row.wrapping_sub(dr as usize);
                    col = col.wrapping_sub(dc as usize);

                    (dr, dc) = Self::turn_right(dr, dc);
                } else {
                    seen.insert((row, col, dr, dc));
                }
            }
        }
        total
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 4696);
        assert_eq!(p2, 1443);
    }
}

fn main() {
    aoc_2024::run_day(6, &Day6);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}